onnxruntime/core/framework/sparse_csrcformat_rep.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/sparse_csrcformat_rep.h"
+#include "core/framework/data_transfer_manager.h"
+
+namespace onnxruntime {
+
+SparseCsrcFormatRep::~SparseCsrcFormatRep() = default;
+
+void SparseCsrcFormatRep::InitIndices(const TensorShape& inner_shape, void* inner_data,
+                 const TensorShape& outer_shape, void* outer_data) {
+  auto ml_type = DataTypeImpl::GetType<int64_t>();
+  inner_indices_ = Tensor(ml_type, inner_shape, inner_data, Location());
+  outer_indices_ = Tensor(ml_type, outer_shape, outer_data, Location());
+}
+
+void SparseCsrcFormatRep::InitIndices(const TensorShape& inner_shape, const TensorShape& outer_shape) {
+  auto ml_ind_type = DataTypeImpl::GetType<int64_t>();
+  void* inner_data = IndicesStart(ml_ind_type->Size());
+  const auto inner_indices_size = inner_shape.Size() * ml_ind_type->Size();
+  void* outer_data = reinterpret_cast<uint8_t*>(inner_data) + inner_indices_size;
+  InitIndices(inner_shape, inner_data, outer_shape, outer_data);
+}
+
+Status SparseCsrcFormatRep::Copy(const DataTransferManager& data_transfer_manager,
+                                 const AllocatorPtr& allocator,
+                                 int exec_q_id, std::unique_ptr<SparseRep>& dst_rep) const {
+  const int64_t required_size = RequiredAllocationSize();
+  auto rep_copy = std::make_unique<SparseCsrcFormatRep>(Values().DataType(),
+                                                        Values().Shape(),
+                                                        required_size,
+                                                        allocator, Major());
+  rep_copy->InitIndices(Inner().Shape(), Outer().Shape());
+  ORT_RETURN_IF_ERROR(data_transfer_manager.CopyTensor(Inner(), rep_copy->MutableInner(), exec_q_id));
+  ORT_RETURN_IF_ERROR(data_transfer_manager.CopyTensor(Outer(), rep_copy->MutableOuter(), exec_q_id));
+  dst_rep = std::move(rep_copy);
+  return Status::OK();
```

Comment:
Given we have one contiguous buffer can we just copy that in one go using the data transfer manager, and handle the metadata about that data (i.e. copying the datatype/shape for the indices Tensor instances and fixing their data pointers) separately (e.g. member function of the SparseRep class could handle all this internally)? 

This would minimize the perf hit if there's a device copy involved, and simplify the interfaces for the different SparseRep classes as they wouldn't need mutable accessors to all their internals.

Reply:
Yes, I am thinking the same.

Reply:
When is all the info not known to construct a SparseTensor rep? Don't you need to know the sizes of everything to construct the single buffer, and if you know the sizes wouldn't you know the data for the indices etc. as well?

Reply:
It is true, that I will know the sizes, but the values and indices are typically reside in different chunks of memory when we first construct the Tensor, so we need to make more than one copy on construction, not to mention that values can be different types, including strings.

Reply:
Copy logic has been reworked.